File: AddOneToList/AddOneToList.cpp
// AddOneToList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

void PrintList(ListNode* pHead)
{
    if (pHead == nullptr)
        return;

    std::cout << pHead->data;
    pHead = pHead->next;

    while (pHead != nullptr)
    {
        std::cout << ", " << pHead->data;
        pHead = pHead->next;
    }
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

bool AddOne(ListNode* head)
{
    if (head == nullptr)
        return false;

    bool resultIsTen = false;
    if (head->next == nullptr)
    {
        head->data++;
        if (head->data == 10)
        {
            resultIsTen = true;
            head->data = 0;
        }
        return resultIsTen;
    }

    if (AddOne(head->next))
    {
        head->data++;
        if (head->data == 10)
        {
            resultIsTen = true;
            head->data = 0;
        }
    }

    return resultIsTen;
}

ListNode* AddOneToList(ListNode* head)
{
    if (AddOne(head))
    {
        ListNode* pNode = new ListNode(1);
        pNode->next = head;
        return pNode;
    }

    return head;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of nodes in list: ";
        std::cin >> count;
        if (count <= 0)
            break;

        ListNode* pHead = nullptr;
        ListNode* pTail = GetListFromInput(&pHead, count);
        
        pHead = AddOneToList(pHead);

        std::cout << "The list after added one: ";
        PrintList(pHead);
        std::cout << std::endl;

        FreeList(pHead);
    }
}

File: Go/AddOneToList.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) *ListNode {
	if len(v) == 0 {
		return nil
	}

	if len(v) == 1 {
		return &ListNode{v[0], nil}
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead
}

func (pNode *ListNode) Print() {
	fmt.Print(pNode.data)
	if pNode.next != nil {
		fmt.Print(", ")
		pNode.next.Print()
	} else {
		fmt.Println("")
	}
}

func AddOneToLinkedList(pNode *ListNode) int {
	if pNode == nil {
		return 1
	}

	if AddOneToLinkedList(pNode.next) == 1 {
		pNode.data++
		if pNode.data == 10 {
			pNode.data = 0
			return 1
		}
	}

	return 0
}

func AddOneToList(head *ListNode) *ListNode {
	n := AddOneToLinkedList(head)
	if n != 0 {
		pNewHead := &ListNode{n, nil}
		pNewHead.next = head
		return pNewHead
	}

	return head
}

func main() {
	for {
		fmt.Print("Number of nodes: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := GetArrayFromInput(count, "Please enter the list: ")
		pHead := GetListFromArray(vec)
		pHead = AddOneToList(pHead)

		fmt.Print("The list after one added: ")
		pHead.Print()
	}
}
